lib/versioning/gradle/compare.ts

```diff
@@ -192,6 +192,24 @@ export function compare(left: string, right: string): number {
   return 0;
 }
 
+export function valid(input: string): Token[] | null {
+  if (!input) {
+    return null;
+  }
+
+  if (!regEx(/^[-._+a-zA-Z0-9]+$/i).test(input)) {
+    return null;
+  }
+
+  if (regEx(/^latest\.?/i).test(input)) {
+    return null;
+  }
+
+  const tokens = tokenize(input);
+
+  return tokens;
+}
+
 export function isVersion(input: string): boolean {
```

Comment:
All functions should use `valid` where appropriate.

Reply:
We use `isVersion` to define functions `isSingleVersion` and `isCompatible`, as they have the same return types. 

And there are some places where only a boolean value is required and `isVersion` is still in use at those places only. We can get boolean from `valid` using `!!` it just didn't look as clean and since `isVerison` was still there I have used it.

Reply:
We should remove this, as it's too much duplicate code. Or reuse `valid` to reduce duplicate code. To have less or changes.

We use `!!` at many other places and it's a valid operation.

Reply:
```js
export function valid(input: string): Token[] | null {
  if (isVersion(input)) {
    return tokenize(input);
  }
  return null;
}
```

Reply:
@viceice Will this do?